components/core/src/clp/streaming_compression/lzma/Compressor.cpp

```diff
@@ -0,0 +1,205 @@
+#include "Compressor.hpp"
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+
+#include <lzma.h>
+#include <spdlog/spdlog.h>
+
+#include "../../ErrorCode.hpp"
+#include "../../FileWriter.hpp"
+#include "../../TraceableException.hpp"
+#include "../../type_utils.hpp"
+
+namespace clp::streaming_compression::lzma {
+auto Compressor::open(FileWriter& file_writer) -> void {
+    if (nullptr != m_compressed_stream_file_writer) {
+        throw OperationFailed(ErrorCode_NotReady, __FILENAME__, __LINE__);
+    }
+
+    m_lzma_stream.detach_input();
+    if (false
+        == m_lzma_stream.attach_output(
+                m_compressed_stream_block_buffer.data(),
+                m_compressed_stream_block_buffer.size()
+        ))
+    {
+        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+    }
+    m_compressed_stream_file_writer = &file_writer;
+    m_uncompressed_stream_pos = 0;
+}
+
+auto Compressor::close() -> void {
+    if (nullptr == m_compressed_stream_file_writer) {
+        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+    }
+
+    if (m_lzma_stream.avail_in() > 0) {
```

Comment:
Looking at how currently our code is written, there isn't a valid case where m_lzma_stream.avail_in() would end-up greater than 0.

I think we should throw an error instead.

Reply:
ok